subprojects/execution/src/test/groovy/org/gradle/internal/execution/steps/ExecutionTest.groovy

```diff
@@ -203,4 +205,32 @@ class ExecutionTest extends Specification {
         }
     }
 
+    def execute(Supplier<UnitOfWork> work) {
+        executionStep.execute(new IncrementalChangesContext() {
+            @Override
+            Optional<ExecutionStateChanges> getChanges() {
```

Comment:
I find it pretty annoying that those things need to be overridden together each time we create a context further down the road, only to pass things through. Do you think that maybe the `rebuildReason` and the execution states should be merged into one object?

Reply:
The execution states each, and the rebuild reason again will be added in a separate `Step`, at least if we keep the original structure (which I intend to do at least first). I'd like to review these choices later, and improve on them once we have all the stuff we want moved over to the execution engine.